File: backend/app/schemas/runs.py
"""Pydantic shapes for the run lifecycle + approvals."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

RunStatus = Literal["pending", "running", "succeeded", "failed"]


class RunListItem(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    name: str
    status: RunStatus
    record_count: int
    match_count: int
    error: str | None
    created_at: datetime
    completed_at: datetime | None


class AuthorityMatchResponse(BaseModel):
    id: uuid.UUID
    control_number: str
    entity_text: str
    entity_kind: str
    role: str
    matched_name: str
    mazal_id: str
    viaf_id: str
    wikidata_qid: str
    confidence: str
    source: str
    payload: dict[str, Any] = Field(default_factory=dict)
    approved: bool
    approved_by: uuid.UUID | None
    approved_at: datetime | None


class RunDetail(RunListItem):
    matches: list[AuthorityMatchResponse]


class ApprovalUpdate(BaseModel):
    approved: bool


class ApprovalBatch(BaseModel):
    match_ids: list[uuid.UUID]
    approved: bool


class RunMarcRecord(BaseModel):
    control_number: str
    marc: dict[str, Any]


class AuthorityMatchEdit(BaseModel):
    """Partial update of an authority match — every field optional."""
    matched_name: str | None = None
    mazal_id: str | None = None
    viaf_id: str | None = None
    wikidata_qid: str | None = None
    confidence: Literal["high", "medium", "low"] | None = None
    role: str | None = None
    entity_text: str | None = None


class RecordEdit(BaseModel):
    """MARC JSON replacement — UI sends the dict, server stores it."""
    marc: dict[str, Any]


class AuthorityAutoApproveRule(BaseModel):
    """Rule for bulk-approving authority candidates by predicate."""
    confidence_levels: list[Literal["high", "medium", "low"]] = Field(
        default_factory=lambda: ["high", "medium", "low"],
    )
    sources: list[str] = Field(default_factory=list)     # empty = any source
    entity_kinds: list[str] = Field(default_factory=list) # empty = any kind
    min_source_count: int = Field(default=1, ge=1, le=4)
    require_ai_pass: bool = False
    respect_ai_fail: bool = True


class AiVerdictResponse(BaseModel):
    """Returned by /runs/{id}/matches/{mid}/ai-verify and embedded in
    payload['ai_verdict'] on the match itself."""

    overall: Literal["full", "partial", "fail", "abstain"]
    reasoning: str
    model: str
    judged_at: str
    fallback: bool = False    # true when we couldn't reach Gemini

File: frontend/src/api/runs.ts
import { api, csrfHeaders } from "@/api/client";

export type RunStatus = "pending" | "running" | "succeeded" | "failed";

export interface RunListItem {
  id: string;
  project_id: string;
  name: string;
  status: RunStatus;
  record_count: number;
  match_count: number;
  error: string | null;
  created_at: string;
  completed_at: string | null;
}

export interface ExistsIn {
  status: "grounded" | "novel" | "wrong_field" | "unknown" | string;
  fields?: string[];
  note?: string;
}

export interface AuthorityMatch {
  id: string;
  control_number: string;
  entity_text: string;
  entity_kind: string;
  role: string;
  matched_name: string;
  mazal_id: string;
  viaf_id: string;
  wikidata_qid: string;
  confidence: "high" | "medium" | "low" | string;
  source: string;
  payload: Record<string, unknown>;
  approved: boolean;
  approved_by: string | null;
  approved_at: string | null;
  exists_in?: ExistsIn | null;
}

export interface RunDetail extends RunListItem {
  matches: AuthorityMatch[];
}

export interface RunMarcRecord {
  control_number: string;
  marc: Record<string, unknown>;
}

export const Runs = {
  listForProject: (projectId: string) =>
    api.get<RunListItem[]>(`/projects/${projectId}/runs`),

  create: async (projectId: string, file: File): Promise<RunListItem> => {
    const fd = new FormData();
    fd.append("file", file);
    const res = await fetch(`/api/projects/${projectId}/runs`, {
      method: "POST",
      credentials: "include",
      headers: { ...csrfHeaders("POST") },
      body: fd,
    });
    if (!res.ok) {
      let detail = res.statusText;
      try {
        const d = (await res.json()) as { detail?: string };
        if (d?.detail) detail = d.detail;
      } catch { /* ignore */ }
      throw new Error(detail);
    }
    return (await res.json()) as RunListItem;
  },

  get: (id: string) => api.get<RunDetail>(`/runs/${id}`),

  listMatches: (id: string) => api.get<AuthorityMatch[]>(`/runs/${id}/matches`),

  listRecords: (id: string) =>
    api.get<string[]>(`/runs/${id}/records`),

  getRecord: (id: string, controlNumber: string) =>
    api.get<RunMarcRecord>(`/runs/${id}/records/${encodeURIComponent(controlNumber)}`),

  editRecord: (runId: string, controlNumber: string, marc: Record<string, unknown>) =>
    api.patch<RunMarcRecord>(
      `/runs/${runId}/records/${encodeURIComponent(controlNumber)}`,
      {marc},
    ),

};
